feat: created generic Codepen Snippet component

ErickPetru opened the pull request:
## Description of Problem
Right now, more than 20 files along the documentation have Codepen embeds with small changes on each one, like this:

``` html
<p class="codepen" data-height="300" data-theme-id="39028" data-default-tab="result" data-user="Vue" data-slug-hash="KKpRVvJ" data-preview="true" data-editable="true" style="height: 300px; box-sizing: border-box; display: flex; align-items: center; justify-content: center; border: 2px solid; margin: 1em 0; padding: 1em;" data-pen-title="Attribute dynamic binding">
  <span>See the Pen <a href="https://codepen.io/team/Vue/pen/KKpRVvJ">
  Attribute dynamic binding</a> by Vue (<a href="https://codepen.io/Vue">@Vue</a>)
  on <a href="https://codepen.io">CodePen</a>.</span>
</p>
<script async src="https://static.codepen.io/assets/embed/ei.js"></script>
```

This large block of copied-and-pasted content takes up a lot of space in each place and is hard to maintain.

Also, as a documentation translation maintainer, I feel how it's annoying to keep all embeds translated in the same way when we have different collaborators translating different files.

And last but not least, using Single File Components inside Markdown with VuePress is so cool! Why not take advantage of this tool one more time to improve this scenario?

## Proposed Solution
A generic Codepen Snippet component, with props covering all different situations where Codepen embeds are use a little different across the documentation, but with good defaults that cover the majority of the use cases, to keep the usage simple.

## Additional Information
I took the liberty of changing all the Codepen embeds currently used in the documentation to use the new component, in order to test its effectiveness (and also to facilitate the work of the core team). I am available to make any additional adjustments to it if necessary.

NataliaTepluhina commented on src/.vuepress/components/common/codepen-snippet.vue:
```diff
@@ -0,0 +1,81 @@
+      default: null,
+    },
+
+    title: {
```
**nitpick**: should we make `title` and `slug` required? This way it would be easier to see that we didn't miss important ones.

NataliaTepluhina reviewed (approved):
@ErickPetru I like the idea! I've left one nitpick to codepen component but overall it looks good to me 👍🏻

skirtles-code commented on src/guide/introduction.md:
```diff
@@ -195,12 +180,7 @@ This example demonstrates that we can bind data to not only text and attributes,
-  on <a href="https://codepen.io">CodePen</a>.</span>
-</p>
-<script async src="https://static.codepen.io/assets/embed/ei.js"></script>
+<common-codepen-snippet title="Conditional rendering" slug="oNXdbpB" />
```
I think this needs `tab="js,result"`.

This particular example encourages the reader to edit the code.

ErickPetru replied:
Yes, the original has `data-default-tab="js,result"`. I just forgot to add with the `tab` prop.

skirtles-code commented on src/guide/transitions-overview.md:
```diff
@@ -180,21 +170,11 @@ If we were to apply these states to a transition, it would look something like t
-  on <a href="https://codepen.io">CodePen</a>.</span>
-</p>
-<script async src="https://static.codepen.io/assets/embed/ei.js"></script>
+<common-codepen-snippet title="Bouncing Ball Demo" slug="zxJWBJ" />
```
There seem to be some changes to this example. Were these intentional?

ErickPetru replied:
Thanks, by mistake I forgot to add the custom author as props in this case.

skirtles-code commented on src/examples/commits.md:
```diff
@@ -2,9 +2,4 @@
-  on <a href="https://codepen.io">CodePen</a>.</span>
-</p>
-<script async src="https://static.codepen.io/assets/embed/ei.js"></script>
+<common-codepen-snippet title="Vue 3 Commits" slug="RwaWmzY" />
```
Does this need `tab="js,result"`?

ErickPetru replied:
Yes, added.

skirtles-code commented on src/.vuepress/components/common/codepen-snippet.vue:
```diff
@@ -0,0 +1,81 @@
+    class="codepen"
+    data-theme-id="39028"
+    data-preview="true"
+    data-editable="true"
```
Some of the current examples are not editable. Specifically, those in the Examples section.

Is changing this intentional?

ErickPetru replied:
Initially I thought we could define a standard to it, but after more thinking I realized that each section can have its own reasons to allow or not inline editable examples. So, my intention is not to change any current behavior, just improve future maintenance with the component. My recent commits added this behavior as props too.

skirtles-code commented on src/.vuepress/components/common/codepen-snippet.vue:
```diff
@@ -0,0 +1,81 @@
+<template>
+  <p
+    class="codepen"
+    data-theme-id="39028"
```
Some of the current examples use a different value here.

I'm not sure what it does but is the change intentional?

ErickPetru replied:
I really think its a problem, the standardized theme should be used along all the docs. But I don't want to change behavior with this PR, so I've created a new `theme` prop too. And it'll be easy to solve the different theme in some files in the future, if the team wants.

skirtles-code commented on src/.vuepress/components/common/codepen-snippet.vue:
```diff
@@ -0,0 +1,81 @@
+  <p
+    class="codepen"
+    data-theme-id="39028"
+    data-preview="true"
```
A lot of the current examples don't have this.

Again, I'm not sure what it does but I was wondering whether the change was intentional?

ErickPetru replied:
Yes, the same case as I said about `editable`. Initially I thought `preview` should be always true. But my intention here isn't changing already established behavior, so I updated all files that hadn't `data-preview` putting the prop `:preview="false"`.

skirtles-code commented on src/.vuepress/components/common/codepen-snippet.vue:
```diff
@@ -0,0 +1,81 @@
+    :data-pen-title="title"
+    :style="`height: ${height}px`">
+    <span>See the Pen <a :href="href">{{ title }}</a>
+    by {{ name ? name : user }} (<a :href="`https://codepen.io/${user}`">@{{user}}</a>)
```
Perhaps replace `name ? name : user` with `name || user`?

ErickPetru replied:
Nice catch, tanks.

skirtles-code commented on src/guide/transitions-overview.md:
```diff
@@ -70,12 +70,7 @@ const Demo = {
-  on <a href="https://codepen.io">CodePen</a>.</span>
-</p>
-<script async src="https://static.codepen.io/assets/embed/ei.js"></script>
+<common-codepen-snippet title="Create animation with a class Lists" slug="ff45b91caf7a98c8c9077ad8ab539260" tab="css,result" />
```
Should the word 'Lists' be included in this title?

ErickPetru replied:
No, another mistake.

skirtles-code commented on src/guide/transitions-enterleave.md:
```diff
@@ -139,12 +134,7 @@ Vue.createApp(Demo).mount('#demo')
-  on <a href="https://codepen.io">CodePen</a>.</span>
-</p>
-<script async src="https://static.codepen.io/assets/embed/ei.js"></script>
+<common-codepen-snippet title="Different Enter and Leave 0dfa7869450ef43d6f7bd99022bc53e2" slug="3466d06fb252a53c5bc0a0edb0f1588a" tab="css,result" />
```
I think there's been a copy/paste error here with the `title` and `slug`.

ErickPetru replied:
For sure, corrected.

skirtles-code commented on src/guide/a11y-basics.md:
```diff
@@ -54,12 +54,7 @@ export default {
 
-<p class="codepen" data-height="350" data-theme-id="light" data-default-tab="js,result" data-user="mlama007" data-slug-hash="VwepxJa" style="height: 350px; box-sizing: border-box; display: flex; align-items: center; justify-content: center; border: 2px solid; margin: 1em 0; padding: 1em;" data-pen-title="Skip to Main">
-  <span>See the Pen <a href="https://codepen.io/mlama007/pen/VwepxJa">
-  Skip to Main</a> by Maria (<a href="https://codepen.io/mlama007">@mlama007</a>)
```
There are a few examples that used to mention 'Maria' that don't seem to be doing that anymore. Is this intentional?

ErickPetru replied:
No, just forgot the `name` prop there and I already updated in the recent commits. But in the future I think the core team could migrate everything to the **team/Vue** Codepen account.

skirtles-code reviewed (changes requested):
A few suggestions from me.

bencodezen reviewed (changes requested):
Thanks for the work on this @ErickPetru! Your refactoring has shown us that there's some clean up that needs to do in terms of standardizing some things so it's easier to turn into a component. I think we should probably hold on this until those changes are made so you can update the PR.

ErickPetru commented:
Hi @bencodezen, my intention with this PR was not to change current behavior.

I really forgot some possible adaptations at the first pass, but after @skirtles-code review, I've already sent more commits with everything (I hope) in the same way it was before.

I believe it's secure to merge after the next review phase. Indeed, even future standardization on the embeds will be a lot easier with the new component. Also, feel free to ask for more changes if you want.

skirtles-code commented:
@ErickPetru Currently there are conflicts with `master`.

Given what @bencodezen said, I'm unclear whether more conflicts are coming or if this is now good to proceed.

I'm happy to do another review but I'd rather wait for confirmation that there aren't more conflicts coming first.

ErickPetru commented:
@skirtles-code and @bencodezen I saw the recent changes on `examples/modal.md` and `examples/grid-component.md` adding the updated Codepen URLs. So I just sent these two updated using the CodepenSnippet component.

skirtles-code commented on src/guide/transitions-state.md:
```diff
@@ -45,25 +45,15 @@ const Demo = {
-  on <a href="https://codepen.io">CodePen</a>.</span>
-</p>
-<script async src="https://static.codepen.io/assets/embed/ei.js"></script>
+<common-codepen-snippet title="Updating SVG" slug="a8e00648d4df6baa1b19fb6c31c8d17e" :height="493" tab="js,result" :editable="false" />
```
The original example had `data-height="500"` but `style="height: 493px;"`.

From checking the deployed version of this branch it seems that `:height="500"` would be a better match for the live docs.

skirtles-code commented on src/cookbook/editable-svg-icons.md:
```diff
@@ -143,7 +143,7 @@ export default {
 We're applying `refs` to the groups of paths we need to move, and as both sides of the scissors have to move in tandem, we'll create a function we can reuse where we'll pass in the `refs`. The use of GreenSock helps resolve animation support and `transform-origin` issues across browser.
 
-<p data-height="300" data-theme-id="0" data-slug-hash="dJRpgY" data-default-tab="result" data-user="Vue" data-embed-version="2" data-pen-title="Editable SVG Icon System: Animated icon" class="codepen">See the Pen <a href="https://codepen.io/team/Vue/pen/dJRpgY/">Editable SVG Icon System: Animated icon</a> by Vue (<a href="https://codepen.io/Vue">@Vue</a>) on <a href="https://codepen.io">CodePen</ ...
+<common-codepen-snippet title="Editable SVG Icon System: Animated icon" slug="dJRpgY" :preview="false" :editable="false" version="2" />
```
The `theme` for this example was previously set to `0`.

skirtles-code commented on src/guide/transitions-state.md:
```diff
@@ -157,5 +142,4 @@ Vue can help. Since SVGs are just data, we only need examples of what these crea
 
-<p data-height="400" data-theme-id="light" data-slug-hash="YZBGNp" data-default-tab="result" data-user="sdras" data-embed-version="2" data-pen-title="Vue-controlled Wall-E" class="codepen">See the Pen <a href="https://codepen.io/sdras/pen/YZBGNp/">Vue-controlled Wall-E</a> by Sarah Drasner (<a href="https://codepen.io/sdras">@sdras</a>) on <a href="https://codepen.io">CodePen</a>.</p>
-<script async src="https://production-assets.codepen.io/assets/embed/ei.js"></script>
+<common-codepen-snippet title="Vue-controlled Wall-E" slug="YZBGNp" :height="400" :team="false" user="sdras" name="Sarah Drasner" :editable="false" :preview="false" version="2" />
```
This was previously using the `light` theme.

skirtles-code reviewed (changes requested):
Some more suggestions from me.

I've not made any subjective decisions about whether things matter, I've just reported anything that seems to have changed from what was there previously.

I agree that this component should make it easier to remove the current inconsistencies.

ErickPetru commented:
Sorry for the delay, hard job weeks here currently.
I've just updated with the corrections after the second @skirtles-code review.

skirtles-code reviewed (changes requested):
There's a new CodePen example at the bottom of `transitions-enterleave.md`.

Usually I'd suggest migrating it post-merge but in this case it seems that the component can't co-exist with non-component CodePens. That new example gets a thick, black border around it as a result.

I did a bit of debugging and it seems to be a consequence of how SSR and the CodePen script interact with each other. So long as everything is using the component it should be fine.

As SSR is involved you can only see the problem by going directly to the right page:

https://deploy-preview-548--vue-docs-next-preview.netlify.app/guide/transitions-enterleave.html#transitioning-between-components

If you navigate there via another page it all behaves.

ErickPetru commented:
Incredible was your catch about SSR with Codepen embeds without the new component, @skirtles-code!

So as you suggested, I just updated `guide/transitions-enterleave.md` changing the last embed to also use the component.

Then I just realized that the project could have direct `<iframe>` embeds to, and I was right. So I changed to use the component also in the files `guide/optimizations.md` and `guide/reactivity.md`. While the way it works becomes a little different (from an direct `<iframe>` to a `<p>` converted to an `<iframe>` by https://static.codepen.io/assets/embed/ei.js), in my tests it offers the same experience to the end user and finally standardize all the guide to use the `CodepenSnippet` component.

Please take a look and talk to me if something more is needed.

bencodezen reviewed (approved):
Thanks for your hard work on this @ErickPetru!
